File: zig/tests/test_adversarial_edge_cases.zig
const std = @import("std");
const testing = std.testing;
const bplustree = @import("bplustree");

test "adversarial: root collapse infinite loop attack" {
    const allocator = testing.allocator;
    
    const Tree = bplustree.BPlusTree(i32, i32);
    var tree = Tree.init(allocator, 4);
    defer tree.deinit();
    
    // Build a multi-level tree
    for (0..64) |i| {
        try tree.insert(@intCast(i), @intCast(i * 10));
    }
    
    // Delete in a pattern that forces repeated root collapses
    var i: i32 = 63;
    while (i >= 0) : (i -= 1) {
        if (@mod(i, 8) != 0) {
            _ = try tree.remove(i);
            // Verify tree is still valid after each deletion
            try testing.expect(tree.len() > 0 or i == 0);
        }
    }
    
    // Tree should now have very few items but still be valid
    var remaining_count: usize = 0;
    var iter = tree.iterator();
    while (iter.next()) |_| {
        remaining_count += 1;
    }
    try testing.expectEqual(tree.len(), remaining_count);
    
    // Try to break it with one more operation
    try tree.insert(100, 1000);
    
    // Verify final state
    try testing.expectEqual(remaining_count + 1, tree.len());
}

test "adversarial: minimum capacity edge cases attack" {
    const allocator = testing.allocator;
    
    // Attack: Use minimum capacity (2) and test all edge cases
    const capacity: usize = 2; // Minimum allowed in Zig implementation
    const Tree = bplustree.BPlusTree(i32, i32);
    var tree = try Tree.initWithValidation(allocator, capacity);
    defer tree.deinit();
    
    // Test 1: Exactly capacity items in root leaf
    for (0..capacity) |i| {
        try tree.insert(@intCast(i), @intCast(i));
    }
    
    // This should trigger first split
    try tree.insert(@intCast(capacity), @intCast(capacity));
    
    // Verify split happened correctly
    try testing.expect(tree.getHeight() > 1);
    
    // Test 2: Clear and test deletion patterns
    tree.clear();
    
    // Insert pattern to create specific structure
    for (0..50) |i| {
        try tree.insert(@intCast(i * 2), @intCast(i * 20));
    }
    
    // Delete to leave each node at minimum
    const delete_pattern = [_]i32{ 2, 6, 10, 14, 18, 22, 26, 30, 34, 38, 42, 46, 50, 54, 58 };
    for (delete_pattern) |key| {
        if (tree.contains(key)) {
            _ = try tree.remove(key);
        }
    }
    
    // Try one more deletion - should trigger rebalancing
    _ = try tree.remove(0);
    
    // Verify tree is still valid
    try testing.expect(tree.len() < 50);
}

test "adversarial: odd capacity arithmetic attack" {
    const allocator = testing.allocator;
    
    // Attack: Use odd capacities to expose integer division bugs
    const odd_capacities = [_]usize{ 3, 5, 7, 9, 11 };
    
    for (odd_capacities) |capacity| {
        const Tree = bplustree.BPlusTree(i32, i32);
        var tree = Tree.init(allocator, capacity);
        defer tree.deinit();
        
        // Fill to exactly trigger splits at boundaries
        const total = capacity * 10;
        for (0..total) |i| {
            try tree.insert(@intCast(i), @intCast(i));
        }
        
        // min_keys calculation for odd numbers
        const min_keys = capacity / 2; // Floor division
        _ = min_keys; // Used for validation
        
        // Delete to exactly min_keys in some nodes
        var deleted: usize = 0;
        var j: i32 = @intCast(total - 1);
        while (j >= 0) : (j -= 1) {
            if (deleted >= capacity * 7) {
                break;
            }
            if (@mod(j, 3) != 0) {
                _ = try tree.remove(j);
                deleted += 1;
            }
        }
        
        // Verify min_keys calculation doesn't break invariants
        try testing.expect(tree.len() == total - deleted);
        
        // Tree should still function correctly
        try tree.insert(-1, -1);
        try testing.expectEqual(@as(i32, -1), tree.get(-1).?);
    }
}

test "adversarial: insert remove same key attack" {
    const allocator = testing.allocator;
    
    const Tree = bplustree.BPlusTree(i32, i32);
    var tree = Tree.init(allocator, 4);
    defer tree.deinit();
    
    // Attack: Rapidly insert/remove same keys
    const attack_key = 42;
    
    // Phase 1: Build some structure
    for (0..20) |i| {
        try tree.insert(@intCast(i), @intCast(i));
    }
    
    // Phase 2: Thrash on one key
    for (0..100) |i| {
        if (i % 2 == 0) {
            try tree.insert(attack_key, @intCast(i));
        } else {
            if (tree.contains(attack_key)) {
                _ = try tree.remove(attack_key);
            }
        }
        
        // Verify tree consistency
        var count: usize = 0;
        var iter = tree.iterator();
        while (iter.next()) |_| {
            count += 1;
        }
        try testing.expectEqual(tree.len(), count);
    }
}

test "adversarial: split merge thrashing attack" {
    const allocator = testing.allocator;
    
    const Tree = bplustree.BPlusTree(i32, i32);
    var tree = Tree.init(allocator, 4);
    defer tree.deinit();
    
    // Attack: Insert/delete at split boundaries repeatedly
    const boundary = 4; // Capacity
    
    for (0..50) |cycle| {
        // Fill to just before split
        for (0..boundary - 1) |i| {
            const key = @as(i32, @intCast(cycle * 100 + i));
            try tree.insert(key, key);
        }
        
        // Trigger split
        const split_key = @as(i32, @intCast(cycle * 100 + boundary));
        try tree.insert(split_key, split_key);
        
        // Delete to trigger potential merge
        for (0..boundary + 1) |i| { // Include the split_key
            const key = @as(i32, @intCast(cycle * 100 + i));
            if (tree.contains(key)) {
                _ = try tree.remove(key);
            }
        }
    }
    
    // Tree should be valid but might be empty or have few items
    try testing.expect(tree.len() < 10);
}

test "adversarial: extreme key ordering attack" {
    const allocator = testing.allocator;
    
    const Tree = bplustree.BPlusTree(i32, i32);
    var tree = Tree.init(allocator, 4);
    defer tree.deinit();
    
    // Attack: Insert keys in pathological order
    const n = 100;
    
    // Pattern 1: Alternating high/low
    for (0..n / 2) |i| {
        try tree.insert(@intCast(i), @intCast(i));
        try tree.insert(@intCast(n - i - 1), @intCast(n - i - 1));
    }
    
    // Pattern 2: Delete middle elements
    for (n / 4..3 * n / 4) |i| {
        _ = try tree.remove(@intCast(i));
    }
    
    // Pattern 3: Re-insert in different order
    var j = @as(i32, @intCast(3 * n / 4 - 1));
    while (j >= @as(i32, @intCast(n / 4))) : (j -= 1) {
        try tree.insert(j, j * 2);
    }
    
    // Verify ordering is maintained
    var last_key: ?i32 = null;
    var iter = tree.iterator();
    while (iter.next()) |entry| {
        if (last_key) |lk| {
            try testing.expect(entry.key > lk);
        }
        last_key = entry.key;
    }
}

test "adversarial: concurrent operation simulation attack" {
    const allocator = testing.allocator;
    
    const Tree = bplustree.BPlusTree(i32, i32);
    var tree = Tree.init(allocator, 6);
    defer tree.deinit();
    
    // Simulate concurrent operations by interleaving
    const ops = [_]enum { Insert, Remove, Get, Range }{ .Insert, .Remove, .Get, .Range };
    const keys = [_]i32{ 10, 20, 30, 40, 50, 15, 25, 35, 45, 55 };
    
    // Build initial tree
    for (keys) |key| {
        try tree.insert(key, key * 10);
    }
    
    // Interleaved operations
    for (0..100) |i| {
        const op = ops[i % ops.len];
        const key = keys[i % keys.len];
        
        switch (op) {
            .Insert => {
                try tree.insert(key + @as(i32, @intCast(i)), @intCast(i));
            },
            .Remove => {
                if (tree.contains(key)) {
                    _ = try tree.remove(key);
                }
            },
            .Get => {
                _ = tree.get(key);
            },
            .Range => {
                var results = std.ArrayList(Tree.Entry).init(allocator);
                defer results.deinit();
                try tree.range(key - 5, key + 5, &results);
            },
        }
    }
    
    // Verify final consistency
    try testing.expect(tree.len() > 0);
}

test "adversarial: memory pressure attack" {
    const allocator = testing.allocator;
    
    // Use large value type to create memory pressure
    const LargeValue = struct {
        data: [256]u8,
        
        fn init(n: u8) @This() {
            var val = @This(){ .data = undefined };
            @memset(&val.data, n);
            return val;
        }
    };
    
    const Tree = bplustree.BPlusTree(i32, LargeValue);
    var tree = Tree.init(allocator, 4);
    defer tree.deinit();
    
    // Insert many large values
    for (0..50) |i| {
        try tree.insert(@intCast(i), LargeValue.init(@truncate(i)));
    }
    
    // Delete half
    for (0..25) |i| {
        _ = try tree.remove(@intCast(i * 2));
    }
    
    // Re-insert with different values
    for (0..25) |i| {
        try tree.insert(@intCast(i * 2), LargeValue.init(@truncate(i + 100)));
    }
    
    // Verify all operations succeeded
    try testing.expectEqual(@as(usize, 50), tree.len());
}

test "adversarial: deep tree attack" {
    const allocator = testing.allocator;
    
    const Tree = bplustree.BPlusTree(i32, i32);
    var tree = Tree.init(allocator, 2); // Minimum capacity for maximum depth
    defer tree.deinit();
    
    // Create very deep tree
    for (0..1000) |i| {
        try tree.insert(@intCast(i), @intCast(i));
    }
    
    // Attack deep paths
    var prng = std.Random.DefaultPrng.init(12345);
    const random = prng.random();
    
    for (0..100) |_| {
        const key = random.intRangeAtMost(i32, 0, 999);
        
        if (random.boolean()) {
            try tree.insert(key, key * 2);
        } else if (tree.contains(key)) {
            _ = try tree.remove(key);
        }
    }
    
    // Verify tree didn't degrade
    try testing.expect(tree.getHeight() < 20); // Reasonable bound for balanced tree
}

test "adversarial: boundary value attack" {
    const allocator = testing.allocator;
    
    const Tree = bplustree.BPlusTree(i32, i32);
    var tree = Tree.init(allocator, 4);
    defer tree.deinit();
    
    // Test all boundary conditions
    const boundaries = [_]struct { cap: usize, split_at: usize }{
        .{ .cap = 2, .split_at = 2 },
        .{ .cap = 3, .split_at = 3 },
        .{ .cap = 4, .split_at = 4 },
        .{ .cap = 5, .split_at = 5 },
    };
    
    for (boundaries) |boundary| {
        tree.clear();
        
        // Fill to exactly split point
        for (0..boundary.split_at) |i| {
            try tree.insert(@intCast(i * 10), @intCast(i));
        }
        
        const height_before = tree.getHeight();
        
        // This should trigger split
        try tree.insert(@intCast(boundary.split_at * 10), @intCast(boundary.split_at));
        
        // Verify split occurred when appropriate
        // Note: Actual split behavior depends on implementation details
        if (boundary.split_at > boundary.cap) {
            // Should have split
            try testing.expect(tree.getHeight() >= height_before);
        }
    }
}

File: c/tests/test_adversarial_edge_cases.c
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <string.h>
#include <limits.h>
#include "../src/bplustree.h"

// Test framework (same as other tests)
static int tests_run = 0;
static int tests_passed = 0;

#define TEST(name) \
    static void test_##name(void); \
    static void run_test_##name(void) { \
        printf("Running test: %s... ", #name); \
        fflush(stdout); \
        test_##name(); \
        tests_passed++; \
        printf("PASSED\n"); \
    } \
    static void test_##name(void)

#define ASSERT(condition) \
    do { \
        if (!(condition)) { \
            printf("FAILED\n"); \
            printf("  Assertion failed: %s\n", #condition); \
            printf("  File: %s, Line: %d\n", __FILE__, __LINE__); \
            exit(1); \
        } \
    } while(0)

#define ASSERT_EQ(expected, actual) \
    do { \
        if ((expected) != (actual)) { \
            printf("FAILED\n"); \
            printf("  Expected: %d, Actual: %d\n", (int)(expected), (int)(actual)); \
            printf("  File: %s, Line: %d\n", __FILE__, __LINE__); \
            exit(1); \
        } \
    } while(0)

#define RUN_TEST(name) \
    do { \
        tests_run++; \
        run_test_##name(); \
    } while(0)

// Test 1: Root collapse infinite loop attack
TEST(adversarial_root_collapse_infinite_loop_attack) {
    bplustree_t* tree = bptree_new(8);
    ASSERT(tree != NULL);
    
    // Insert many keys to create a multi-level tree
    for (int i = 0; i < 64; i++) {
        bptree_result_t result = bptree_insert(tree, i, i * 100);
        ASSERT_EQ(BPTREE_OK, result);
    }
    
    ASSERT_EQ(64, bptree_size(tree));
    
    // Delete in a pattern that forces repeated root collapses
    for (int i = 63; i >= 0; i--) {
        if (i % 8 != 0) {
            bptree_result_t result = bptree_remove(tree, i);
            ASSERT_EQ(BPTREE_OK, result);
        }
    }
    
    // Tree should still be functional
    ASSERT_EQ(8, bptree_size(tree));
    
    // Verify remaining keys
    for (int i = 0; i < 64; i += 8) {
        ASSERT(bptree_contains(tree, i));
        int value;
        bptree_result_t result = bptree_get(tree, i, &value);
        ASSERT_EQ(BPTREE_OK, result);
        ASSERT_EQ(i * 100, value);
    }
    
    bptree_free(tree);
}

// Test 2: Minimum capacity edge cases
TEST(adversarial_minimum_capacity_edge_cases) {
    // Test with minimum possible capacity
    bplustree_t* tree = bptree_new(BPTREE_MIN_CAPACITY);
    ASSERT(tree != NULL);
    
    // Insert exactly enough keys to trigger multiple levels
    int max_keys = BPTREE_MIN_CAPACITY * BPTREE_MIN_CAPACITY + BPTREE_MIN_CAPACITY;
    
    for (int i = 0; i < max_keys; i++) {
        bptree_result_t result = bptree_insert(tree, i, i * 2);
        ASSERT_EQ(BPTREE_OK, result);
    }
    
    ASSERT_EQ(max_keys, bptree_size(tree));
    
    // Test iteration in order
    bptree_iterator_t* iter = bptree_iterator_new(tree);
    ASSERT(iter != NULL);
    
    int count = 0;
    int expected_key = 0;
    while (bptree_iterator_has_next(iter)) {
        bptree_entry_t entry;
        bptree_result_t result = bptree_iterator_next(iter, &entry);
        ASSERT_EQ(BPTREE_OK, result);
        ASSERT_EQ(expected_key, entry.key);
        ASSERT_EQ(expected_key * 2, entry.value);
        expected_key++;
        count++;
    }
    
    ASSERT_EQ(max_keys, count);
    bptree_iterator_free(iter);
    bptree_free(tree);
}

// Test 3: Odd capacity arithmetic edge cases
TEST(adversarial_odd_capacity_arithmetic_edge_cases) {
    // Test with odd capacity to stress arithmetic
    bplustree_t* tree = bptree_new(7); // Odd capacity
    ASSERT(tree != NULL);
    
    // Insert in reverse order to stress rebalancing
    for (int i = 100; i >= 1; i--) {
        bptree_result_t result = bptree_insert(tree, i, i * 3);
        ASSERT_EQ(BPTREE_OK, result);
    }
    
    ASSERT_EQ(100, bptree_size(tree));
    
    // Delete every third element to create gaps
    for (int i = 3; i <= 100; i += 3) {
        bptree_result_t result = bptree_remove(tree, i);
        ASSERT_EQ(BPTREE_OK, result);
    }
    
    size_t remaining = 100 - 33; // 33 elements divisible by 3
    ASSERT_EQ(remaining, bptree_size(tree));
    
    // Verify iteration still works correctly
    bptree_iterator_t* iter = bptree_iterator_new(tree);
    ASSERT(iter != NULL);
    
    int count = 0;
    int last_key = 0;
    while (bptree_iterator_has_next(iter)) {
        bptree_entry_t entry;
        bptree_result_t result = bptree_iterator_next(iter, &entry);
        ASSERT_EQ(BPTREE_OK, result);
        
        // Keys should be in ascending order
        ASSERT(entry.key > last_key);
        last_key = entry.key;
        
        // Key should not be divisible by 3
        ASSERT(entry.key % 3 != 0);
        
        count++;
    }
    
    ASSERT_EQ(remaining, count);
    bptree_iterator_free(iter);
    bptree_free(tree);
}

// Test 4: Sequential key exhaustion patterns
TEST(adversarial_sequential_key_exhaustion_patterns) {
    bplustree_t* tree = bptree_new(5);
    ASSERT(tree != NULL);
    
    // Pattern 1: Insert ascending, delete descending
    for (int i = 1; i <= 50; i++) {
        bptree_result_t result = bptree_insert(tree, i, i * 10);
        ASSERT_EQ(BPTREE_OK, result);
    }
    
    for (int i = 50; i >= 26; i--) {
        bptree_result_t result = bptree_remove(tree, i);
        ASSERT_EQ(BPTREE_OK, result);
    }
    
    ASSERT_EQ(25, bptree_size(tree));
    
    // Pattern 2: Insert gaps, then fill them
    for (int i = 51; i <= 75; i += 2) {
        bptree_result_t result = bptree_insert(tree, i, i * 10);
        ASSERT_EQ(BPTREE_OK, result);
    }
    
    for (int i = 52; i <= 74; i += 2) {
        bptree_result_t result = bptree_insert(tree, i, i * 10);
        ASSERT_EQ(BPTREE_OK, result);
    }
    
    // Verify all keys 1-25 and 51-75 exist
    for (int i = 1; i <= 25; i++) {
        ASSERT(bptree_contains(tree, i));
    }
    for (int i = 51; i <= 75; i++) {
        ASSERT(bptree_contains(tree, i));
    }
    
    ASSERT_EQ(50, bptree_size(tree));
    bptree_free(tree);
}

// Test 5: Extreme boundary value insertions
TEST(adversarial_extreme_boundary_value_insertions) {
    bplustree_t* tree = bptree_new(4);
    ASSERT(tree != NULL);
    
    // Test with extreme integer values
    int extreme_keys[] = {
        INT_MIN,
        INT_MIN + 1,
        -1000000,
        -1,
        0,
        1,
        1000000,
        INT_MAX - 1,
        INT_MAX
    };
    int num_keys = sizeof(extreme_keys) / sizeof(extreme_keys[0]);
    
    // Insert in random order
    int insert_order[] = {4, 8, 0, 6, 2, 7, 1, 5, 3};
    
    for (int i = 0; i < num_keys; i++) {
        int key = extreme_keys[insert_order[i]];
        bptree_result_t result = bptree_insert(tree, key, key / 2);
        ASSERT_EQ(BPTREE_OK, result);
    }
    
    ASSERT_EQ(num_keys, bptree_size(tree));
    
    // Verify iteration returns keys in sorted order
    bptree_iterator_t* iter = bptree_iterator_new(tree);
    ASSERT(iter != NULL);
    
    int count = 0;
    int last_key = INT_MIN;
    bool first = true;
    
    while (bptree_iterator_has_next(iter)) {
        bptree_entry_t entry;
        bptree_result_t result = bptree_iterator_next(iter, &entry);
        ASSERT_EQ(BPTREE_OK, result);
        
        if (!first) {
            ASSERT(entry.key > last_key);
        }
        last_key = entry.key;
        first = false;
        count++;
    }
    
    ASSERT_EQ(num_keys, count);
    bptree_iterator_free(iter);
    bptree_free(tree);
}

// Test 6: Pathological split cascade patterns
TEST(adversarial_pathological_split_cascade_patterns) {
    bplustree_t* tree = bptree_new(4);
    ASSERT(tree != NULL);
    
    // Create a scenario that forces cascading splits
    // Insert keys that will cause maximum split cascading
    for (int round = 0; round < 5; round++) {
        int base = round * 100;
        
        // Insert in a pattern that maximizes split propagation
        for (int i = 0; i < 20; i++) {
            int key = base + (i * 13) % 20; // Pseudo-random pattern
            bptree_result_t result = bptree_insert(tree, key, key * 7);
            ASSERT_EQ(BPTREE_OK, result);
        }
    }
    
    ASSERT_EQ(100, bptree_size(tree));
    
    // Verify all keys are accessible
    for (int round = 0; round < 5; round++) {
        int base = round * 100;
        for (int i = 0; i < 20; i++) {
            int key = base + (i * 13) % 20;
            int value;
            bptree_result_t result = bptree_get(tree, key, &value);
            ASSERT_EQ(BPTREE_OK, result);
            ASSERT_EQ(key * 7, value);
        }
    }
    
    bptree_free(tree);
}

// Test 7: Iterator invalidation stress test
TEST(adversarial_iterator_invalidation_stress_test) {
    bplustree_t* tree = bptree_new(6);
    ASSERT(tree != NULL);
    
    // Insert initial data
    for (int i = 0; i < 30; i++) {
        bptree_result_t result = bptree_insert(tree, i, i * 5);
        ASSERT_EQ(BPTREE_OK, result);
    }
    
    // Create iterator and partially consume it
    bptree_iterator_t* iter = bptree_iterator_new(tree);
    ASSERT(iter != NULL);
    
    // Consume first 10 entries
    for (int i = 0; i < 10; i++) {
        ASSERT(bptree_iterator_has_next(iter));
        bptree_entry_t entry;
        bptree_result_t result = bptree_iterator_next(iter, &entry);
        ASSERT_EQ(BPTREE_OK, result);
        ASSERT_EQ(i, entry.key);
    }
    
    // Modify tree structure (should not crash iterator)
    for (int i = 100; i < 120; i++) {
        bptree_result_t result = bptree_insert(tree, i, i * 5);
        ASSERT_EQ(BPTREE_OK, result);
    }
    
    // Continue iteration (may skip some elements due to modifications)
    int remaining_count = 0;
    while (bptree_iterator_has_next(iter)) {
        bptree_entry_t entry;
        bptree_result_t result = bptree_iterator_next(iter, &entry);
        ASSERT_EQ(BPTREE_OK, result);
        remaining_count++;
    }
    
    // Should have seen some remaining elements
    ASSERT(remaining_count > 0);
    
    bptree_iterator_free(iter);
    bptree_free(tree);
}

// Test 8: Memory pressure simulation
TEST(adversarial_memory_pressure_simulation) {
    // Create many small trees to stress memory allocation
    const int num_trees = 100;
    bplustree_t* trees[num_trees];
    
    // Create trees
    for (int t = 0; t < num_trees; t++) {
        trees[t] = bptree_new(4 + (t % 8)); // Varying capacities
        ASSERT(trees[t] != NULL);
        
        // Add some data to each tree
        for (int i = 0; i < 10; i++) {
            int key = t * 1000 + i;
            bptree_result_t result = bptree_insert(trees[t], key, key * 2);
            ASSERT_EQ(BPTREE_OK, result);
        }
    }
    
    // Randomly access trees to stress memory
    for (int access = 0; access < 1000; access++) {
        int tree_idx = access % num_trees;
        int key = tree_idx * 1000 + (access % 10);
        
        int value;
        bptree_result_t result = bptree_get(trees[tree_idx], key, &value);
        ASSERT_EQ(BPTREE_OK, result);
        ASSERT_EQ(key * 2, value);
    }
    
    // Clean up
    for (int t = 0; t < num_trees; t++) {
        bptree_free(trees[t]);
    }
}

// Test 9: Duplicate key update patterns
TEST(adversarial_duplicate_key_update_patterns) {
    bplustree_t* tree = bptree_new(5);
    ASSERT(tree != NULL);
    
    // Insert initial keys
    for (int i = 0; i < 20; i++) {
        bptree_result_t result = bptree_insert(tree, i, i);
        ASSERT_EQ(BPTREE_OK, result);
    }
    
    // Pattern 1: Update every key multiple times
    for (int round = 1; round <= 5; round++) {
        for (int i = 0; i < 20; i++) {
            bptree_result_t result = bptree_insert(tree, i, i * round);
            ASSERT_EQ(BPTREE_OK, result);
        }
        
        // Size should remain constant
        ASSERT_EQ(20, bptree_size(tree));
        
        // Verify values were updated
        for (int i = 0; i < 20; i++) {
            int value;
            bptree_result_t result = bptree_get(tree, i, &value);
            ASSERT_EQ(BPTREE_OK, result);
            ASSERT_EQ(i * round, value);
        }
    }
    
    bptree_free(tree);
}

// Test 10: Range query boundary attack
TEST(adversarial_range_query_boundary_attack) {
    bplustree_t* tree = bptree_new(6);
    ASSERT(tree != NULL);
    
    // Insert keys with gaps to stress range queries
    for (int i = 0; i < 100; i += 5) {
        bptree_result_t result = bptree_insert(tree, i, i * 10);
        ASSERT_EQ(BPTREE_OK, result);
    }
    
    // Test edge case ranges
    struct {
        int start, end;
        int expected_count;
    } test_cases[] = {
        {0, 5, 1},      // Single element
        {0, 6, 2},      // Two elements
        {-10, 0, 0},    // Before range
        {100, 110, 0},  // After range
        {-10, 110, 20}, // Entire range
        {47, 53, 2},    // Mid-range
        {48, 52, 1},    // Between elements
    };
    
    int num_cases = sizeof(test_cases) / sizeof(test_cases[0]);
    
    for (int tc = 0; tc < num_cases; tc++) {
        bptree_iterator_t* iter = bptree_range_iterator_new(tree, 
            test_cases[tc].start, test_cases[tc].end);
        ASSERT(iter != NULL);
        
        int count = 0;
        while (bptree_iterator_has_next(iter)) {
            bptree_entry_t entry;
            bptree_result_t result = bptree_iterator_next(iter, &entry);
            ASSERT_EQ(BPTREE_OK, result);
            
            // Verify key is in range
            ASSERT(entry.key >= test_cases[tc].start);
            ASSERT(entry.key < test_cases[tc].end);
            count++;
        }
        
        ASSERT_EQ(test_cases[tc].expected_count, count);
        bptree_iterator_free(iter);
    }
    
    bptree_free(tree);
}

int main(void) {
    printf("=== B+ Tree C Implementation - Adversarial Edge Cases Tests ===\n\n");
    
    RUN_TEST(adversarial_root_collapse_infinite_loop_attack);
    RUN_TEST(adversarial_minimum_capacity_edge_cases);
    RUN_TEST(adversarial_odd_capacity_arithmetic_edge_cases);
    RUN_TEST(adversarial_sequential_key_exhaustion_patterns);
    RUN_TEST(adversarial_extreme_boundary_value_insertions);
    RUN_TEST(adversarial_pathological_split_cascade_patterns);
    RUN_TEST(adversarial_iterator_invalidation_stress_test);
    RUN_TEST(adversarial_memory_pressure_simulation);
    RUN_TEST(adversarial_duplicate_key_update_patterns);
    RUN_TEST(adversarial_range_query_boundary_attack);
    
    printf("\n=== Test Summary ===\n");
    printf("Tests run: %d\n", tests_run);
    printf("Tests passed: %d\n", tests_passed);
    
    if (tests_passed == tests_run) {
        printf("🎉 All tests passed!\n");
        return 0;
    } else {
        printf("❌ Some tests failed!\n");
        return 1;
    }
}